pandas/core/case_when.py

```diff
@@ -0,0 +1,210 @@
+    if len(bool_length) > 1:
+        raise ValueError("All boolean conditions should have the same length.")
+    bool_length = conditions[0].size
+    if not is_scalar(default):
+        if not isinstance(default, ABCDataFrame):
+            default = convert_to_list_like(default)
+        if not hasattr(default, "shape"):
+            default = pd_array(default, copy=False)
+        if default.ndim > 1:
+            raise ValueError(
+                "The provided default argument should "
+                "either be a scalar or a 1-D array."
+            )
+        if default.size != conditions[0].size:
+            raise ValueError(
+                "The length of the default argument, "
+                "if it is array-like, should "
+                "be the same as the length of any "
+                "of the boolean conditions."
+            )
+    replacements = []
+    common_dtype = []
+    # ideally we could skip these checks and let `Series.mask`
+    # handle it - however, this step is necessary to get
+    # a common dtype for multiple conditions and replacements.
+    for num, replacement in zip(counter, args[-1::-2]):
+        if not is_scalar(replacement):
+            if not isinstance(replacement, ABCDataFrame):
+                replacement = convert_to_list_like(replacement)
+            if not hasattr(replacement, "shape"):
+                replacement = pd_array(replacement, copy=False)
+            if replacement.ndim > 1:
+                raise ValueError(f"replacement{num} should be a 1-D array.")
+            if replacement.size != bool_length:
+                raise ValueError(
+                    f"The size of replacement{num} array"
+                    f"does not match the size of condition{num} array."
+                )
+        arg_dtype, _ = infer_dtype_from(replacement)
+        common_dtype.append(arg_dtype)
+        replacements.append(replacement)
+    if default is not None:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtype.append(arg_dtype)
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            arr
+            if is_scalar(arr)
+            else arr.astype(common_dtype, copy=False)
+            if isinstance(arr, ABCSeries)
+            else pd_array(arr, dtype=common_dtype, copy=False)
+            for arr in replacements
+        ]
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtype, copy=False)
+    else:
+        common_dtype = common_dtype[0]
+    if not isinstance(default, ABCSeries):
+        default_index = range(bool_length)
+        for replacement in replacements[::-1]:
+            if isinstance(replacement, ABCSeries):
+                default_index = replacement.index
+                break
+        default = Series(default, index=default_index, dtype=common_dtype)
```

Comment:
Assuming we respect alignment, why the reverse iteration here and using replacements instead of conditions? If we always use the first condition (it has to have an index), I think that is a more natural choice: it agrees with the idea of iteratively refining the values of the first condition/value pair.

Reply:
No solid reasoning on this one. I'll defer to your suggestion